File: zeroday-server/src/main/kotlin/com/zeroday/model/Faction.kt
package com.zeroday.model

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable
data class Faction(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val tag: String,
    val description: String = "",
    val leaderId: String,
    val members: MutableList<String> = mutableListOf(),
    val joinRequests: MutableList<String> = mutableListOf(),
    val level: Int = 1,
    val experience: Long = 0,
    val mainframe: FactionMainframe = FactionMainframe(),
    val createdAt: Long = System.currentTimeMillis(),
    val isOpen: Boolean = true
)

@Serializable
data class FactionMainframe(
    var level: Int = 1,
    var cpuTotal: Long = 1000,
    var cpuUsed: Long = 0,
    var ramTotal: Long = 4096,
    var ramUsed: Long = 0,
    var bandwidthTotal: Long = 500,
    var bandwidthUsed: Long = 0,
    var creditsPool: Long = 0,
    var securityLevel: Int = 1,
    var passiveIncomeRate: Double = 1.0,
    var activeBuffs: MutableList<FactionBuff> = mutableListOf(),
    var unlockedFeatures: MutableList<String> = mutableListOf("faction_chat"),
    var upgradeProgress: MutableMap<String, Long> = mutableMapOf()
)

@Serializable
data class FactionBuff(
    val name: String,
    val description: String,
    val type: BuffType,
    val value: Double,
    val durationMs: Long,
    val expiresAt: Long = System.currentTimeMillis() + durationMs
)

@Serializable
enum class BuffType {
    EXPERIENCE_MULTIPLIER,
    CREDIT_MULTIPLIER,
    CPU_REGEN_BOOST,
    RAM_REGEN_BOOST,
    SCAN_RANGE_BOOST,
    FIREWALL_BOOST
}

@Serializable
data class FactionUpgrade(
    val id: String,
    val name: String,
    val description: String,
    val category: UpgradeCategory,
    val levels: List<UpgradeLevel>
)

@Serializable
data class UpgradeLevel(
    val level: Int,
    val costCredits: Long,
    val costCpu: Long,
    val costRam: Long,
    val benefit: String,
    val value: Double,
    val requiredFactionLevel: Int = 1
)

@Serializable
enum class UpgradeCategory {
    CPU_CAPACITY,
    RAM_CAPACITY,
    BANDWIDTH_CAPACITY,
    SECURITY,
    PASSIVE_INCOME,
    BUFF_DURATION,
    MEMBER_LIMIT,
    FEATURE_UNLOCK
}

object FactionUpgradesRegistry {
    val allUpgrades: List<FactionUpgrade> = listOf(
        FactionUpgrade("cpu_cap", "CPU Core Cluster", "Expand mainframe processing power",
            UpgradeCategory.CPU_CAPACITY, listOf(
                UpgradeLevel(1, 5000, 0, 0, "+5000 CPU capacity", 5000.0),
                UpgradeLevel(2, 15000, 0, 0, "+10000 CPU capacity", 10000.0, 2),
                UpgradeLevel(3, 50000, 0, 0, "+25000 CPU capacity", 25000.0, 4),
                UpgradeLevel(4, 150000, 0, 0, "+50000 CPU capacity", 50000.0, 6)
            )),
        FactionUpgrade("ram_cap", "Memory Banks", "Upgrade faction RAM pool",
            UpgradeCategory.RAM_CAPACITY, listOf(
                UpgradeLevel(1, 5000, 0, 0, "+8GB RAM", 8192.0),
                UpgradeLevel(2, 15000, 0, 0, "+16GB RAM", 16384.0, 2),
                UpgradeLevel(3, 50000, 0, 0, "+32GB RAM", 32768.0, 4),
                UpgradeLevel(4, 150000, 0, 0, "+64GB RAM", 65536.0, 6)
            )),
        FactionUpgrade("bandwidth_cap", "Fiber Optic Link", "Increase faction bandwidth",
            UpgradeCategory.BANDWIDTH_CAPACITY, listOf(
                UpgradeLevel(1, 3000, 0, 0, "+500 Mbps", 500.0),
                UpgradeLevel(2, 10000, 0, 0, "+1000 Mbps", 1000.0, 2),
                UpgradeLevel(3, 30000, 0, 0, "+2500 Mbps", 2500.0, 4),
                UpgradeLevel(4, 100000, 0, 0, "+5000 Mbps", 5000.0, 6)
            )),
        FactionUpgrade("security", "ICE Security Suite", "Strengthen mainframe defenses",
            UpgradeCategory.SECURITY, listOf(
                UpgradeLevel(1, 2000, 1000, 512, "Security Level +1", 1.0),
                UpgradeLevel(2, 8000, 3000, 1024, "Security Level +2", 2.0, 2),
                UpgradeLevel(3, 25000, 10000, 4096, "Security Level +3", 3.0, 4),
                UpgradeLevel(4, 80000, 30000, 16384, "Security Level +5", 5.0, 6)
            )),
        FactionUpgrade("passive_income", "Automated Mining Rig", "Generate passive credits",
            UpgradeCategory.PASSIVE_INCOME, listOf(
                UpgradeLevel(1, 5000, 2000, 1024, "+2 credits/min per member", 2.0),
                UpgradeLevel(2, 20000, 8000, 4096, "+5 credits/min per member", 5.0, 3),
                UpgradeLevel(3, 60000, 25000, 16384, "+10 credits/min per member", 10.0, 5)
            )),
        FactionUpgrade("member_limit", "Member Expansion", "Increase faction member cap",
            UpgradeCategory.MEMBER_LIMIT, listOf(
                UpgradeLevel(1, 10000, 0, 0, "Max members: 15", 15.0),
                UpgradeLevel(2, 30000, 0, 0, "Max members: 25", 25.0, 3),
                UpgradeLevel(3, 100000, 0, 0, "Max members: 50", 50.0, 5)
            )),
        FactionUpgrade("feature_unlock", "Advanced Modules", "Unlock exclusive faction features",
            UpgradeCategory.FEATURE_UNLOCK, listOf(
                UpgradeLevel(1, 20000, 10000, 4096, "Unlocks: faction_scan", 1.0, 2),
                UpgradeLevel(2, 50000, 25000, 8192, "Unlocks: mass_ddos", 2.0, 4),
                UpgradeLevel(3, 150000, 50000, 32768, "Unlocks: faction_shield", 3.0, 6)
            ))
    )
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Faction.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

public class Faction
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("tag")]
    public string Tag { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("leader_id")]
    public string LeaderId { get; set; } = "";

    [JsonPropertyName("members")]
    public List<string> Members { get; set; } = new();

    [JsonPropertyName("join_requests")]
    public List<string> JoinRequests { get; set; } = new();

    [JsonPropertyName("level")]
    public int Level { get; set; } = 1;

    [JsonPropertyName("experience")]
    public long Experience { get; set; } = 0;

    [JsonPropertyName("mainframe")]
    public FactionMainframe Mainframe { get; set; } = new();

    [JsonPropertyName("created_at")]
    public long CreatedAt { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    [JsonPropertyName("is_open")]
    public bool IsOpen { get; set; } = true;
}

public class FactionMainframe
{
    [JsonPropertyName("level")]
    public int Level { get; set; } = 1;

    [JsonPropertyName("cpu_total")]
    public long CpuTotal { get; set; } = 1000;

    [JsonPropertyName("cpu_used")]
    public long CpuUsed { get; set; } = 0;

    [JsonPropertyName("ram_total")]
    public long RamTotal { get; set; } = 4096;

    [JsonPropertyName("ram_used")]
    public long RamUsed { get; set; } = 0;

    [JsonPropertyName("bandwidth_total")]
    public long BandwidthTotal { get; set; } = 500;

    [JsonPropertyName("bandwidth_used")]
    public long BandwidthUsed { get; set; } = 0;

    [JsonPropertyName("credits_pool")]
    public long CreditsPool { get; set; } = 0;

    [JsonPropertyName("security_level")]
    public int SecurityLevel { get; set; } = 1;

    [JsonPropertyName("passive_income_rate")]
    public double PassiveIncomeRate { get; set; } = 1.0;

    [JsonPropertyName("active_buffs")]
    public List<FactionBuff> ActiveBuffs { get; set; } = new();

    [JsonPropertyName("unlocked_features")]
    public List<string> UnlockedFeatures { get; set; } = new() { "faction_chat" };

    [JsonPropertyName("upgrade_progress")]
    public Dictionary<string, long> UpgradeProgress { get; set; } = new();
}

public class FactionBuff
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("type")]
    public BuffType Type { get; set; }

    [JsonPropertyName("value")]
    public double Value { get; set; }

    [JsonPropertyName("duration_ms")]
    public long DurationMs { get; set; }

    [JsonPropertyName("expires_at")]
    public long ExpiresAt { get; set; }
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum BuffType
{
    [JsonPropertyName("experience_multiplier")]
    EXPERIENCE_MULTIPLIER,
    [JsonPropertyName("credit_multiplier")]
    CREDIT_MULTIPLIER,
    [JsonPropertyName("cpu_regen_boost")]
    CPU_REGEN_BOOST,
    [JsonPropertyName("ram_regen_boost")]
    RAM_REGEN_BOOST,
    [JsonPropertyName("scan_range_boost")]
    SCAN_RANGE_BOOST,
    [JsonPropertyName("firewall_boost")]
    FIREWALL_BOOST
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum UpgradeCategory
{
    [JsonPropertyName("cpu_capacity")]
    CPU_CAPACITY,
    [JsonPropertyName("ram_capacity")]
    RAM_CAPACITY,
    [JsonPropertyName("bandwidth_capacity")]
    BANDWIDTH_CAPACITY,
    [JsonPropertyName("security")]
    SECURITY,
    [JsonPropertyName("passive_income")]
    PASSIVE_INCOME,
    [JsonPropertyName("buff_duration")]
    BUFF_DURATION,
    [JsonPropertyName("member_limit")]
    MEMBER_LIMIT,
    [JsonPropertyName("feature_unlock")]
    FEATURE_UNLOCK
}

public class FactionUpgrade
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("category")]
    public UpgradeCategory Category { get; set; }

    [JsonPropertyName("levels")]
    public List<UpgradeLevel> Levels { get; set; } = new();
}

public class UpgradeLevel
{
    [JsonPropertyName("level")]
    public int Level { get; set; }

    [JsonPropertyName("cost_credits")]
    public long CostCredits { get; set; }

    [JsonPropertyName("cost_cpu")]
    public long CostCpu { get; set; }

    [JsonPropertyName("cost_ram")]
    public long CostRam { get; set; }

    [JsonPropertyName("benefit")]
    public string Benefit { get; set; } = "";

    [JsonPropertyName("value")]
    public double Value { get; set; }

    [JsonPropertyName("required_faction_level")]
    public int RequiredFactionLevel { get; set; } = 1;
}

public static class FactionUpgradesRegistry
{
    public static List<FactionUpgrade> AllUpgrades { get; } = new()
    {
        new FactionUpgrade
        {
            Id = "cpu_cap", Name = "CPU Core Cluster", Description = "Expand mainframe processing power",
            Category = UpgradeCategory.CPU_CAPACITY,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 5000, Benefit = "+5000 CPU capacity", Value = 5000.0 },
                new UpgradeLevel { Level = 2, CostCredits = 15000, Benefit = "+10000 CPU capacity", Value = 10000.0, RequiredFactionLevel = 2 },
                new UpgradeLevel { Level = 3, CostCredits = 50000, Benefit = "+25000 CPU capacity", Value = 25000.0, RequiredFactionLevel = 4 },
                new UpgradeLevel { Level = 4, CostCredits = 150000, Benefit = "+50000 CPU capacity", Value = 50000.0, RequiredFactionLevel = 6 }
            }
        },
        new FactionUpgrade
        {
            Id = "ram_cap", Name = "Memory Banks", Description = "Upgrade faction RAM pool",
            Category = UpgradeCategory.RAM_CAPACITY,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 5000, Benefit = "+8GB RAM", Value = 8192.0 },
                new UpgradeLevel { Level = 2, CostCredits = 15000, Benefit = "+16GB RAM", Value = 16384.0, RequiredFactionLevel = 2 },
                new UpgradeLevel { Level = 3, CostCredits = 50000, Benefit = "+32GB RAM", Value = 32768.0, RequiredFactionLevel = 4 },
                new UpgradeLevel { Level = 4, CostCredits = 150000, Benefit = "+64GB RAM", Value = 65536.0, RequiredFactionLevel = 6 }
            }
        },
        new FactionUpgrade
        {
            Id = "bandwidth_cap", Name = "Fiber Optic Link", Description = "Increase faction bandwidth",
            Category = UpgradeCategory.BANDWIDTH_CAPACITY,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 3000, Benefit = "+500 Mbps", Value = 500.0 },
                new UpgradeLevel { Level = 2, CostCredits = 10000, Benefit = "+1000 Mbps", Value = 1000.0, RequiredFactionLevel = 2 },
                new UpgradeLevel { Level = 3, CostCredits = 30000, Benefit = "+2500 Mbps", Value = 2500.0, RequiredFactionLevel = 4 },
                new UpgradeLevel { Level = 4, CostCredits = 100000, Benefit = "+5000 Mbps", Value = 5000.0, RequiredFactionLevel = 6 }
            }
        },
        new FactionUpgrade
        {
            Id = "security", Name = "ICE Security Suite", Description = "Strengthen mainframe defenses",
            Category = UpgradeCategory.SECURITY,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 2000, CostCpu = 1000, CostRam = 512, Benefit = "Security Level +1", Value = 1.0 },
                new UpgradeLevel { Level = 2, CostCredits = 8000, CostCpu = 3000, CostRam = 1024, Benefit = "Security Level +2", Value = 2.0, RequiredFactionLevel = 2 },
                new UpgradeLevel { Level = 3, CostCredits = 25000, CostCpu = 10000, CostRam = 4096, Benefit = "Security Level +3", Value = 3.0, RequiredFactionLevel = 4 },
                new UpgradeLevel { Level = 4, CostCredits = 80000, CostCpu = 30000, CostRam = 16384, Benefit = "Security Level +5", Value = 5.0, RequiredFactionLevel = 6 }
            }
        },
        new FactionUpgrade
        {
            Id = "passive_income", Name = "Automated Mining Rig", Description = "Generate passive credits",
            Category = UpgradeCategory.PASSIVE_INCOME,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 5000, CostCpu = 2000, CostRam = 1024, Benefit = "+2 credits/min per member", Value = 2.0 },
                new UpgradeLevel { Level = 2, CostCredits = 20000, CostCpu = 8000, CostRam = 4096, Benefit = "+5 credits/min per member", Value = 5.0, RequiredFactionLevel = 3 },
                new UpgradeLevel { Level = 3, CostCredits = 60000, CostCpu = 25000, CostRam = 16384, Benefit = "+10 credits/min per member", Value = 10.0, RequiredFactionLevel = 5 }
            }
        },
        new FactionUpgrade
        {
            Id = "member_limit", Name = "Member Expansion", Description = "Increase faction member cap",
            Category = UpgradeCategory.MEMBER_LIMIT,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 10000, Benefit = "Max members: 15", Value = 15.0 },
                new UpgradeLevel { Level = 2, CostCredits = 30000, Benefit = "Max members: 25", Value = 25.0, RequiredFactionLevel = 3 },
                new UpgradeLevel { Level = 3, CostCredits = 100000, Benefit = "Max members: 50", Value = 50.0, RequiredFactionLevel = 5 }
            }
        },
        new FactionUpgrade
        {
            Id = "feature_unlock", Name = "Advanced Modules", Description = "Unlock exclusive faction features",
            Category = UpgradeCategory.FEATURE_UNLOCK,
            Levels = new()
            {
                new UpgradeLevel { Level = 1, CostCredits = 20000, CostCpu = 10000, CostRam = 4096, Benefit = "Unlocks: faction_scan", Value = 1.0, RequiredFactionLevel = 2 },
                new UpgradeLevel { Level = 2, CostCredits = 50000, CostCpu = 25000, CostRam = 8192, Benefit = "Unlocks: mass_ddos", Value = 2.0, RequiredFactionLevel = 4 },
                new UpgradeLevel { Level = 3, CostCredits = 150000, CostCpu = 50000, CostRam = 32768, Benefit = "Unlocks: faction_shield", Value = 3.0, RequiredFactionLevel = 6 }
            }
        }
    };
}
